### PR TITLE
[macOS] Adding QuickLook feature

### PR DESCRIPTION
**Pull request by MatthiasWM**
- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones

_The code is located in `src/MacAppBundle/QuickLook` so it does not interfere with other platforms or macOS command line builds_

- [x]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum

_This PR is in relation to [Forum: New feature: Quicklook for macOS](https://forum.freecadweb.org/viewtopic.php?f=10&t=71803)_

- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [x]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`

_This commit has no impact on the actual app. Also I will need hep at some point to build a full app on macOS Monterey._

- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

_Will do._

### CODE REVIEW

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [640548795](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/640548795) was triggered at [cb5bc5b](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/cb5bc5b388c217f73c861c64bf3372e32ced81ea). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by MatthiasWM** [comment]
This Pull Request adds macOS QuickLook previews and icons to FreeCAD.

The patch in _Info.plist_ makes *.FCStd files known as filetype `org.freecad.fcstd` to the system. The actual QL plugin source code is located in _src/MacAppBundle/QuickLook_ and is installed in _FreeCAD.app/Contents/Library/QuickLook_ using CMake `install` in the file _src/MacAppBundle/CMakeLists.txt_.

FreeCAD needs to be built as a full macOS app (`brew install --with-macos-app freecad@0.20.1`) and the installed in the `/Applications` folder. A restart of _Finder_ may be required to get QL icons.

**Comment by luzpaz** [comment]
@ipatch if you have any time, would you be available to test this PR ?

**Comment by ipatch** [comment]
> @ipatch if you have any time, would you be available to test this PR ?

sure, give me a few.

**Comment by MatthiasWM** [comment]
Looking at the CI output, I saw that the `QuickLook/main.c` source code is checked on Linux, not finding includes that only exist on macOS (CoreFoundations, etc.).

I did fix the tab and space errors that I found in the branch, so they can be pulled. The tab errors in `main.c` and both `Info.plist` files are generated by Apple tools. If this helps, I'll be happy to fix that manually, but at least for `Info.plist`, they will reappear as soon as those files are edited again.

Anything els I need to fix, please let me know,

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [641547865](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/641547865) was triggered at [a95b59b](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/a95b59b099e2f8ae7ef34fc76722697f6cbfe463). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by ipatch** [comment]
@MatthiasWM 

per the github web ui recommendations for testing out this PR, i ran the below commands on my local build box (same box i use to build the bottles for the homebrew freecad tap). i using the macos mojave vm because it gives me the least amount friction when building an app bundle ie. a `FreeCAD.app`

```
git remote add MatthiasWM https://github.com/MatthiasWM/FreeCAD.git
git fetch MatthiasWM ql3
git switch --track MatthiasWM/ql3
```

i pushed your PR to fork of freecad so you can see if might have messed something up, but so far from running those cmds on my setup cmake is giving me `configuration incomplete` message using the below cmake command

```
cmake \
-DHOMEBREW_PREFIX=$bp \
-DPYTHON_EXECUTABLE="$bp/opt/python@3.10/bin/python3" \
-DPYTHON_INCLUDE_DIR="$bp/opt/python@3.10/Frameworks/Python.framework/Headers" \
-DCMAKE_PREFIX_PATH="$pthmedfile;$pthswig;$pthqt;$pthshiboken2;$pthpyside2;$pthvtk82;$pthocc;$pthicu4c" \
-DCMAKE_BUILD_TYPE=RelWithDebInfo \
-DBUILD_FEM_NETGEN=0 \
-DFREECAD_USE_3DCONNEXION=0 \
-DBUILD_QT5=1 \
-DBUILD_ENABLE_CXX_STD=C++17 \
-Wno-dev \
-DFREECAD_CREATE_MAC_APP=1 \
-DCMAKE_INSTALL_PREFIX="$code/freecad-git/installs/$(basename $PWD)" \
-DBUILD_ARCH=1 \
-DBUILD_DRAFT=1 \
-DBUILD_TUX=1 \
-DBUILD_OPENSCAD=1 \
-DBUILD_TECHDRAW=1 \
-DBUILD_SPREADSHEET=1 \
-DBUILD_DRAWING=1 \
\
../../freecad-src/
```

obviously the local shell variables are just that, local for my particular build setup on this box. however i'm seeing the below output after running that cmake command in my build dir. output has been trimmed. can provide full output of the cmake command if needed.

```
=================================================
Now run 'cmake --build /opt/code/freecad-git/builds/build.release.0.20.1.pr7491' to build FreeCAD
=================================================

-- Configuring incomplete, errors occurred!
See also "/opt/code/freecad-git/builds/build.release.0.20.1.pr7491/CMakeFiles/CMakeOutput.log".
See also "/opt/code/freecad-git/builds/build.release.0.20.1.pr7491/CMakeFiles/CMakeError.log".
```

and probably where cmake begins to fail,

```
-- Performing Test _flag_found - Failed
CMake Error at src/MacAppBundle/CMakeLists.txt:37 (string):
  string begin index: -1 is out of range 0 - 74


CMake Error at src/MacAppBundle/CMakeLists.txt:38 (string):
  string sub-command REGEX, mode MATCHALL needs at least 5 arguments total to
  command.


CMake Error at src/MacAppBundle/CMakeLists.txt:39 (string):
  string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command.


CMake Error at src/MacAppBundle/CMakeLists.txt:40 (string):
  string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command.


CMake Error at src/MacAppBundle/CMakeLists.txt:37 (string):
  string begin index: -1 is out of range 0 - 76


CMake Error at src/MacAppBundle/CMakeLists.txt:38 (string):
  string sub-command REGEX, mode MATCHALL needs at least 5 arguments total to
  command.


CMake Error at src/MacAppBundle/CMakeLists.txt:39 (string):
  string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command.


CMake Error at src/MacAppBundle/CMakeLists.txt:40 (string):
  string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command.


-- Checking for module 'icu-uc'
--   Found icu-uc, version 71.1
find: /usr/local/Cellar/nglib: No such file or directory
Quicklook: copy framework template from <</opt/code/freecad-git/freecad-src/src/MacAppBundle>>/QuickLook/ to <</opt/code/freecad-git/installs/build.release.0.20.1.pr7491/FreeCAD.app/Contents>>/Library/QuickLook
--
==============
Summary report
==============
```

i hope this helps. and let me know if i need to amend or update my PR branch at some point.

cheers

**Comment by MatthiasWM** [comment]
@ipatch Thanks for trying my code. I have not touched lines 37 to 40 in MacAppBundle/CMakeLists.txt. My changes are in lines 128 to 144. The lines handle some Python code location fixes and are unrelated to my changes.

I have generally trouble compiling FreeCAD into a bundle. I am using `brew install -v --HEAD --keep-tmp --with-macos-app mcad@0.20.1` where `mcad@0.20.1` is a modified version of `freecad@0.20.1`, pointing to my repo and branch instead of freecad.master . I was never able to use cmake directly. The resulting .app bundle is bad and can not be moved without breaking. I'd love to know how the weekly (daily) M1 builds are actually made.

Not sure if this is related at all, but some CMake parts expect HomeBrew to be in /usr/local which are in /opt/homebrew/ in a fresh homeprew setup: [Homebrew Patch](https://github.com/FreeCAD/homebrew-freecad/blob/master/patches/freecad-0.20.1-macos-app-bundling.patchl)

**Comment by ipatch** [comment]
> @ipatch Thanks for trying my code. I have not touched lines 37 to 40 in MacAppBundle/CMakeLists.txt. My changes are in lines 128 to 144. The lines handle some Python code location fixes and are unrelated to my changes.
> 

thanks. yeah i just popped this tab open again to update my response. indeed those lines are related to collecting / gathering the `.pth` files contained with the python3.x site-packages dir and adding those additional `site-packages` ie. shiboken2 and pyside2 and possible matplotlib if installed, and others as glob looks for `homebrew*` related `.pth` files. i had manually created a `.pth` file both shiboken2 and pyside2 and sure enough i just renamed files for temp purposes and the cmake step finishes without errors. 👍

all that brings another issue though, i think these checks should be a tad bit more robust, but that's separate issue for another day. as my manually added `.pth` files only contain a directory path such as the one below,

```
/Users/brewmaster/homebrew/opt/shiboken2@5.15.5/lib/python3.10/site-packages
```

> I have generally trouble compiling FreeCAD into a bundle. I am using `brew install -v --HEAD --keep-tmp --with-macos-app mcad@0.20.1` where `mcad@0.20.1` is a modified version of `freecad@0.20.1`, pointing to my repo and branch instead of freecad.master . I was never able to use cmake directly. The resulting .app bundle is bad and can not be moved without breaking. I'd love to know how the weekly (daily) M1 builds are actually made.
> 

i believe the weekly builds are made using conda. which is something we should look into. as it looks this plugin might only be applicable for homebrew builds freecad which would be unfortunate for those who are using the conda build and would enjoy using this plugin. me personally i'm not familiar with the conda build system so you'll have to ping somebody who is familiar with that stuff.

> Not sure if this is related at all, but some CMake parts expect HomeBrew to be in /usr/local which are in /opt/homebrew/ in a fresh homeprew setup: [Homebrew Patch](https://github.com/FreeCAD/homebrew-freecad/blob/master/patches/freecad-0.20.1-macos-app-bundling.patchl?rgh-link-date=2022-09-15T20%3A10%3A14Z)

actually the freecad source was setup a while ago where somebody (too lazy, to look through the git blame at the moment) but they added a useful cmake var where you add a `HOMEBREW_PREFIX` as you've pointed out, i believe the default `$bp` (thats my short abbr for `HOMEBREW_PREFIX`), points to `/usr/local` for the default install of homebrew on intel where as the default install path on an m1 box i believe is in the `/opt/homebrew` as you've probably found out. now homebrew does support custom install paths, as you probably read in my cmake command. this useful for a multiuser system, especially so where the github CI only works with a homebrew install in `/usr/local` for intel and the entire setup is tore down after each run (oooffff 😖). and you've also figured out the `--keep-tmp` flag for brew for not having go through arduous process over and over again. kudos.

---

all that said 😥, i'll finish the cmake configure, then run through `make` `make install` and upload the app bundle to the homebrew-freecad if everything works out. and thanks for making this plugin. i enjoy using quicklook myself from time to time. 🎉

**Comment by MatthiasWM** [comment]
> all that brings another issue though, i think these checks should be a tad bit more robust, but that's separate issue for another day. as my manually added `.pth` files only contain a directory path such as the one below,
> 
> ```
> /Users/brewmaster/homebrew/opt/shiboken2@5.15.5/lib/python3.10/site-packages
> ```

Well, at least we get stuck at the same places. Same happens to me if I use cmake directly, but not if I use homebrew. So even with `--keep-tmp`, I still have to rebuild from scratch and through github, but at least I can see what cmake generates.
 
> i believe the weekly builds are made using conda. which is something we should look into. as it looks this plugin might only be applicable for homebrew builds freecad which would be unfortunate for those who are using the conda build and would enjoy using this plugin. me personally i'm not familiar with the conda build system so you'll have to ping somebody who is familiar with that stuff.

I was missing files in the repo to even try condo, but then again, I know nothing about condo. I'll see if I find someone who does. Conda does build a bundle, and I hope (assumed) that it is based on cmake as well, so QuickLook would be part of the release. I'll follow up on that.

**Comment by MatthiasWM** [comment]
Ok, I get FreeCAD to install via Conda, but it does not seem helpful for development. It's not building from source, AFAICT, and I can't even find the binary (much less a bundle). Here is the process, following a bunch of online docs (using bash):
```
brew install anaconda
find /opt/homebrew -name conda
# and add that to PATH, here it's /opt/homebrew/anaconda3/condabin/conda
conda init bash
# and relaunch bash
conda config --add channels conda-forge
conda create --name fcenv-dev --channel freecad/label/dev freecad
conda activate fcenv-dev
FreeCAD # launches an almost complete version of FreeCAD
which FreeCAD
# /opt/homebrew/anaconda3/envs/fcenv-dev/bin/FreeCAD
ls -asl /opt/homebrew/anaconda3/envs/fcenv-dev/bin/FreeCAD
# 152 -rwxrwxr-x  2 matt  staff  77008 Sep 14 15:19 /opt/homebrew/anaconda3/envs/fcenv-dev/bin/FreeCAD
ls -asl /opt/homebrew/anaconda3/envs/fcenv-dev/bin/
# No file named FreeCAD found
```

**Comment by MatthiasWM** [comment]
You are right, the Conda script does indeed include its own Info.plist wich does not have the changes to register the file type. Sigh.

[Script](https://github.com/FreeCAD/FreeCAD-Bundle/blob/master/conda/osx-arm64/create_bundle.sh)
[Info.plist template](https://github.com/FreeCAD/FreeCAD-Bundle/blob/master/conda/osx-arm64/Info.plist.template)

**Comment by ipatch** [comment]
your PR build finished. 👍

ran `make install` got the usual error message i've seen the past.

```
-- Installing: /opt/code/freecad-git/installs/pr7491/FreeCAD.app/Contents/Library/QuickLook/QuicklookFCStd.qlgenerator/Contents/Info.plist
-- Installing: /opt/code/freecad-git/installs/pr7491/FreeCAD.app/Contents/Library/QuickLook/QuicklookFCStd.qlgenerator/Contents/MacOS/QuicklookFCStd
-- Making bundle relocatable...
-- INFO: Analyzing bundle dependencies...
-- ERROR: Unable to find LC_DYLD_LOAD entry: @rpath/libc++abi.1.dylib
-- ERROR: Failed to resolve dependency in /Users/brewmaster/homebrew/opt/llvm/lib/libc++.1.dylib
Traceback (most recent call last):
  File "/opt/code/freecad-git/freecad-src/src/Tools/MakeMacBundleRelocatable.py", line 391, in <module>
    main()
  File "/opt/code/freecad-git/freecad-src/src/Tools/MakeMacBundleRelocatable.py", line 374, in main
    build_deps_graph(graph, bundle_path, dir_filter, search_paths)
  File "/opt/code/freecad-git/freecad-src/src/Tools/MakeMacBundleRelocatable.py", line 241, in build_deps_graph
    deps = create_dep_nodes(list_install_names(k2), s_paths)
  File "/opt/code/freecad-git/freecad-src/src/Tools/MakeMacBundleRelocatable.py", line 166, in create_dep_nodes
    raise LibraryNotFound(lib_name + " not found in given search paths")
__main__.LibraryNotFound: libc++abi.1.dylib not found in given search paths
-- Installing: /opt/code/freecad-git/installs/pr7491/FreeCAD.app/Contents/share/License.txt
-- Installing: /
```

which is unrelated to your PR.

however, is there post installation step that is required after building the `FreeCAD.app` bundle in order to get the quicklook feature working?

so i'm seeing this using quicklook with finder

image

**Comment by MatthiasWM** [comment]
The bundle must be moved to `/Applications`. This should be enough for macOS to pick up on the file type. You can check that by running 

```
qlmanage -m | grep freecad
  org.freecad.fcstd -> /Applications/FreeCAD.app/Contents/Library/QuickLook/QuicklookFCStd.qlgenerator (0)
```

If it does not, use Finder to throw the bundle in the Trash and then move it back into `/Applications`. If it still has not picked up on the new type, a reboot may help. If it still isn't there, `/Applications/FreeCAD.app/Contents/Info.plist` wasn't installed correctly by the CMake script.

If the manager was found, you may need to restart Finder (Option-Rightclick, and select "Relaunch" in the menu).

If all that still fails, you can at least check with `qlmanage -t some.FCStd` and `qlmanage -p some.FCStd` if it created thumbnails and previews. 

Oh, and not all .FCStd files actually have an embedded preview file. Older versions of FreeCAD did not generate thise, and some versions just generated the FC logo as a thumbnail (unzip the FCStd file and search for `thumbnails/Thumbnail.png`). My code will not launch FreeCAD to actually generate a thumbnail if there is none in the file.

**Comment by MatthiasWM** [comment]
**SORRY**, **SORRY**, **SORRY** - I pushed code via git that contains the merge from the master branch, and I do not know how to undo this! Please don't apply the PR in this state.

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [641904435](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/641904435) was triggered at [fbd46a0](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/fbd46a0bc65358c77797f6816389c3493495ba9e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by looooo** [comment]
> You are right, the Conda script does indeed include its own Info.plist wich does not have the changes to register the file type.
> 
> [Script](https://github.com/FreeCAD/FreeCAD-Bundle/blob/master/conda/osx-arm64/create_bundle.sh) [Info.plist template](https://github.com/FreeCAD/FreeCAD-Bundle/blob/master/conda/osx-arm64/Info.plist.template)
> 
> @looooo Hi Lorenz. Would it make sense and be possible to take the Info.plist from the main repo and also include my patches? I am not familiar with Condo, but the CMake setup has the basics to build app bundles and make QuickLook work.

I am not sure if this works. For the conda-based bundles the build and bundle process are two separate tasks. Are there any conditions for the Info.plist to be build?

**Comment by ipatch** [comment]
@MatthiasWM 

finished rebuilding your PR, and produced an app bundle. 👍

- copied the bundle to the `/Applications` dir.

ran the the below commands.

```
brewmaster@vmmojave ~/D/freecad.docs> qlmanage -t bs1.FCStd
Testing Quick Look thumbnails with files:
        bs1.FCStd
2022-09-16 15:42:44.288 qlmanage[40630:14060285] *** CFMessagePort: bootstrap_register(): failed 1100 (0x44c) 'Permission denied', port = 0x8b1f, name = 'com.apple.tsm.portname'
See /usr/include/servers/bootstrap_defs.h for the error codes.
* No thumbnail created for /Users/brewmaster/Documents/freecad.docs/bs1.FCStd
Done producing thumbnails
```

---

i'm seeing the below in my mojave vm.

image


---

do you mind attaching a file that does contain a thumbnail to this PR via a comment. and i can download that file and test?

**Comment by ipatch** [comment]
attached is a file I created with your app bundle `FreeCAD.ap` using this PR. I'm still not seeing a quicklook preview.

[bss1.FCStd.zip](https://github.com/FreeCAD/FreeCAD/files/9587261/bss1.FCStd.zip)

---

below is a screen grab of what the file looks like,

image

**Comment by ipatch** [comment]
> > all that brings another issue though, i think these checks should be a tad bit more robust, but that's separate issue for another day. as my manually added `.pth` files only contain a directory path such as the one below,
> > ```
> > /Users/brewmaster/homebrew/opt/shiboken2@5.15.5/lib/python3.10/site-packages
> > ```
> 
> Well, at least we get stuck at the same places. Same happens to me if I use cmake directly, but not if I use homebrew. So even with `--keep-tmp`, I still have to rebuild from scratch and through github, but at least I can see what cmake generates.
> 
> > i believe the weekly builds are made using conda. which is something we should look into. as it looks this plugin might only be applicable for homebrew builds freecad which would be unfortunate for those who are using the conda build and would enjoy using this plugin. me personally i'm not familiar with the conda build system so you'll have to ping somebody who is familiar with that stuff.
> 
> I was missing files in the repo to even try conda, but then again, I know nothing about condo. I'll see if I find someone who does. Conda does build a bundle, and I hope (assumed) that it is based on cmake as well, so QuickLook would be part of the release. I'll follow up on that.

all that above should probably be opened in a separate issue, but add a link here for completeness sake,

https://github.com/FreeCAD/FreeCAD/commit/a2127e0f0089d7e3a53b8235fd16ba4b41ae9e07#r84213038

**Comment by MatthiasWM** [comment]
@ipatch The file you attached has no Thumbnail.png. You can enable this feature using the FreeCAD preferences. I attached a file that does have the thumbnail (just rename the FCStd file to something.zip and unzip it. Inside should be a folder named "thumbnails", containing a file named "Thumbnail.png").

Reminder to self: final output should be sized using the png size; output image could contain a note stating there was no Thumbnail; can we suggest that the user changes the preferences? ; should we start the entire FreeCAD app in the background to generate the Thumbnail.png (not good IMHO as FC will be launched for each preview or icon)?

[test8.FCStd.zip](https://github.com/FreeCAD/FreeCAD/files/9592398/test8.FCStd.zip) -> rename to test8.FCStd to preview, or to test8.zip to see contents.

Screenshot 2022-09-17 at 13 19 09

**Comment by ipatch** [comment]
that same document i added previously after turning on the option you mentioned above.


image

**Comment by MatthiasWM** [comment]
That's how it is supposed to look. Great!

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [645847973](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/645847973) was triggered at [eb2017b](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/eb2017bf38fd4bd6306c909e164c41deb48bbe59). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by MatthiasWM** [comment]
@looooo I added a PR [136](https://github.com/FreeCAD/FreeCAD-Bundle/pull/136) to FreeCAD-Bundle, so QuickLook should work with Conda bundles as well.

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [647973705](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/647973705) was triggered at [7dfabea](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/7dfabeadde7ac56ea7249883d7b71e53ff7a6b3d). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by MatthiasWM** [comment]
Suspended! I will have to do some updates to make this work with Conda.

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [652251812](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/652251812) was triggered at [8112d6e](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/8112d6ead0bca8a86fd6e12a08ce64ed934ba485). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [655575115](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/655575115) was triggered at [8ea2888](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/8ea2888b15294b21e50708956519286165f043e7). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by MatthiasWM** [comment]
The last commit modifies the CMake scripts to build the QL plugin for Conda/Mamba as well (before it was limited to Homebrew).  This also requires the corresponding PR in FreeCAD/FreeCAD-Bundle. This has been tested and works on M1 Monterey.

The QL plugin is unsigned, just like the app itself. To activate QL, the `FreeCAD.app` must be launched at least once after being moved into `/Applications`. Previews (select in Finder and press the spacebar) should work right away. Icons work after relaunching Finder, or logging out and back in again, or reboot.

**Comment by luzpaz** [comment]
Is this ready for merge?

**Comment by MatthiasWM** [comment]
I believe that it is ready. All my tests worked for my M1 machine. It should work on Intel machines too.

**Comment by luzpaz** [comment]
Shall we merge and see what macos testers say? I could also ping looo to trigger an appimage build so we can have more testers

**Comment by MatthiasWM** [comment]
Yes please. It's important to know that the preferences by default will not generate thumbnails, so Quicklook will show the default icon on those FCStd files. Other than that, Fingers crossed...

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [674769567](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/674769567) was triggered at [7005d4b](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/7005d4b958cc3a4f6e18d1b392eddbbbe4b4aa17). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by MatthiasWM** [comment]
Sorry about the last second addition. This should allow script icons on Homebrew as well.

**Comment by luzpaz** [comment]
Is this ready to be reviewed ?

**Comment by MatthiasWM** [comment]
Yes, it's ready

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [674776242](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/674776242) was triggered at [ac8fe0c](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/ac8fe0c1eafe9a894f0e6f71b65033de45fe1660). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/MacAppBundle/CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,27 @@
+
+#
+# Build and install QuickLook for FCStd files (org.freecad.fcstd)
+# This is used by Homebrew and Conda/Mamabe scripts alike,
+#
+if(FREECAD_CREATE_MAC_APP OR (APPLE AND BUID_WITH_CONDA))
```
Comment (luzpaz) [inline src/MacAppBundle/CMakeLists.txt:6]: `BUID` or `BUILD` ?

Comment (luzpaz) [reply src/MacAppBundle/CMakeLists.txt:6]: Also, line 4 looks like it should be `Mamba` not `Mamabe`

Comment (MatthiasWM) [reply src/MacAppBundle/CMakeLists.txt:6]: Sigh. BUILD_WITH_CONDA it is. Sliped testing thanks to not clearing artefacts correctly.

Comment (MatthiasWM) [reply src/MacAppBundle/CMakeLists.txt]: Fixed.

Comment (luzpaz) [reply src/MacAppBundle/CMakeLists.txt]: See https://github.com/FreeCAD/FreeCAD/pull/7491#discussion_r1002753490

Comment (MatthiasWM) [reply src/MacAppBundle/CMakeLists.txt]: Typo also fixed.

Comment (MatthiasWM) [reply src/MacAppBundle/CMakeLists.txt]: #7491 It's confined to a single issue that needs to be implemented in two redundant build systems. If you prefer, I'll be happy to separate this into three PRs. It's in the history anyway as I implemented Homebrew first and then just added the Conda/Mamba stuff, which is just line 6.

**Comment by luzpaz** [comment]
@ipatch can you re-test and review ?

@looooo there is https://github.com/FreeCAD/FreeCAD-Bundle/pull/136 can we merge and have the macOS folk test it ?  (bump the appimage weekly release to make that happen?)

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [674782892](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/674782892) was triggered at [b39ab4d](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/b39ab4dc563f691c07f68675ac9017c7fca8fed9). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
@ipatch sorry for bugging you...any chance you can test this ?

**Comment by ipatch** [comment]
> @ipatch sorry for bugging you...any chance you can test this ?

sure give me a few.

**Comment by ipatch** [comment]
i was able to build locally using my macos mojave vm. 👍

🎃

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491">pipeline status</a> for feature branch [PR_7491](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7491). Pipeline [681965791](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/681965791) was triggered at [cbc6fcb](https://github.com/FreeCAD/FreeCAD/pull/7491/commits/cbc6fcbd5023830c443bbce935086711b79ea3e6). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by chennes** [comment]
Since @ipatch reports success with this plugin above, I'm going to merge it in so we can get more OS X users in the loop, and can catch anything it's missing. This is a tough one to test!

**Comment by ipatch** [comment]
@chennes no need to revert any changes. i'm going to update the patch file for the freecad formula file in the homebrew-freecad tap to work with this recent merge.

**Comment by MatthiasWM** [comment]
Thanks!

**Comment by luzpaz** [comment]
Thank you @MatthiasWM !

**Pull request merged**